src/order/succ_pred/basic.lean

```diff
@@ -976,3 +998,86 @@ lemma pred.rec_top (p : α → Prop) (htop : p ⊤) (hpred : ∀ a, p a → p (p
 pred.rec htop (λ x _ h, hpred x h) (le_top : a ≤ ⊤)
 
 end order_top
+
+namespace fin
+
+instance : ∀ {n : ℕ}, succ_order (fin n)
+| 0 := by constructor; exact elim0
+| (n+1) :=
+{ succ := λ i, if i < fin.last n then i + 1 else i,
+  le_succ := λ a,
+  begin
+    cases n,
+    { exact (subsingleton.elim _ _).le },
+    split_ifs,
+    { rw [le_iff_coe_le_coe, coe_add_one_of_lt h],
+      exact (a : ℕ).le_succ },
+    { exact le_rfl }
+   end,
+  maximal_of_succ_le := λ a,
+  begin
+    split_ifs,
+    { rw [le_iff_coe_le_coe, coe_add_one_of_lt h],
+      intro h,
+      have := antisymm (a : ℕ).le_succ h,
+      exact absurd this (lt_add_one _).ne },
+    { simp [eq_last_of_not_lt h, le_last] }
+  end,
+  succ_le_of_lt := λ a b h, let h := h.trans_le (le_last _) in
+                            by rwa [if_pos h, le_iff_coe_le_coe, coe_add_one_of_lt h],
+  le_of_lt_succ := λ a b h,
+  begin
+    split_ifs at h with hab,
+    swap, { exact h.le },
+    rw [lt_iff_coe_lt_coe, coe_add_one_of_lt hab] at h,
+    exact nat.le_of_lt_succ h
+  end }
+
+@[simp] lemma succ_eq {n : ℕ} :
+succ_order.succ = λ a, if a < fin.last n then a + 1 else a := rfl
+
+lemma succ_apply {n : ℕ} (a : fin (n + 1)) :
+succ_order.succ a = if a < fin.last n then a + 1 else a := rfl
+
+/-- By sending `x` to `last n - x`, `fin n` is order-equivalent to its `order_dual`. -/
+def equiv_order_dual : ∀ {n}, fin n ≃o order_dual (fin n)
```

Comment:
This is definitely not in the right file. Also, maybe worth swaping the RHS and LHS and calling it `order_iso.dual_fin`?

Reply:
yeah, I know; I'll move it later, but I was wondering if we could transfer `succ_order` to `pred_order` (and vice versa) using this? I'll change the naming when I move it

Reply:
I think it's much cleaner to define the instance by hand. Then `pred` will be something like `if 0 < a then a - 1 else 0` instead of `n.last - if n.last - a < n.last then n.last - a + 1 else n.last`.

Reply:
flipping the sides of this order-iso by default is quite annoying, so do you think it'll be fine if we just `.symm` it? (because then `map_rel_iff'` takes in two terms of `order_dual (fin n)` as opposed to `fin n`)

Reply:
My point is that the more complicated stuff should go on the LHS.